Fix (and test) codegen issues w/ types that conflict with Rust types inc `Result`

anelson opened the pull request:
I ran into oxidecomputer/typify#568 when trying to use typify to generate Rust bindings for the [Model Context
Protocol](https://spec.modelcontextprotocol.io). Anthropic have published a [JSON Schema specification for this protocol](https://github.com/modelcontextprotocol/specification/blob/main/schema/schema.json), but as soon as I tried to codegen Rust bindings for it, I ran into the aforementioned issue.

I added the entire MCP JSON schema to `typify/tests/schemas/model-context-protocol.json`, which as expected immediately caused the `schemas` test to fail.  Fixing this was a relatively simple matter of finding every place in the codegen that referred to the Rust `std::result::Result` type as simply `Result`, and modify the codegen to use the fully-qualified `::std::result::Result` type name instead.. That was enough to make the MCP schema work right.

But then I looked a bit closer and realized there were some other Rust built-in types that were being referenced in such a way as to cause either errors during codegen or generation of code that doesn't compile.

So I made a diabolical test schema `rust-collisions.json` that is truly cursed.  It defines a bunch of types that collide with Rust built-in types and keywords, and made the `schemas` test explode with such force that there may have been inquiries from the neighbors.  Upon reassuring them that that's just how the Rust compiler expresses affection, I made some more codegen changes to resolve all of the issues that my cursed schema exposed, by using fully-qualified type paths `::std::....` for every built-in Rust type that is referenced in the generated code.

I've tried to make the changes minimal, and to my knowledge none of the changes I have made would be breaking from the perspective of any user of the generated Rust code.

I had to modify several existing tests in `typify-impl` that broke because they were expecting the non-canonicalized versions of Rust standard types.  Pay particular attention to the [hack I put in `test_util::validate_output_impl`](https://github.com/anelson/typify/blob/e6be4468dcd88f3ae1c856132ead425080124bbb/typify-impl/src/test_util.rs#L95-L152) which is not something I'm particularly proud of but I think it retains the spirit of the test.

Closes #568

ahl commented on typify-impl/src/structs.rs:
```diff
@@ -373,12 +373,12 @@ pub(crate) fn generate_serde_attr(
         (StructPropertyState::Optional, TypeEntryDetails::Option(_)) => {
             serde_options.push(quote! { default });
-            serde_options.push(quote! { skip_serializing_if = "Option::is_none" });
+            serde_options.push(quote! { skip_serializing_if = "std::option::Option::is_none" });
```
Should we prefix with :: ?

anelson replied:
Good catch, yes we should.  There's also a `std::vec::Vec` right below it which also needs to be fully qualified.

ahl commented on typify-impl/Cargo.toml:
```diff
@@ -17,7 +17,7 @@ schemars = "0.8.21"
 serde = "1.0.215"
 serde_json = "1.0.133"
-syn = { version = "2.0.89", features = ["full"] }
+syn = { version = "2.0.89", features = ["full", "visit-mut"] }
```
Is this new feature only needed for tests?

anelson replied:
Yes, good catch, I moved that feature to the `dev-dependencies`.

ahl commented on typify/tests/schemas.rs:
```diff
@@ -23,6 +23,8 @@ fn test_schemas() {
 }
 
 fn validate_schema(path: std::path::PathBuf) -> Result<(), Box<dyn Error>> {
+    println!("Testing the processing of schema: {}", path.display());
```
Did you mean to leave this?

ahl replied:
If it was useful feel free to leave it

ahl commented on typify-impl/src/test_util.rs:
```diff
@@ -89,6 +92,57 @@ fn validate_output_impl<T: JsonSchema + Schema>(ignore_variant_names: bool) {
+                && !path.segments.is_empty()
+                && path.segments[0].ident == "std"
+            {
+                // Fun additional hack to keep you on your toes:
```
Can you check the history on that? It might have been during a previous round of prefixing

anelson replied:
Yes you are right, it was done in #647.

The tests that have the fully-qualified type paths in them don't seem wrong or irrelevant.  It's not a very common way to declare Rust types outside of proc macros and other codegen, but it's perfectly valid.  

I dislike the fact that a special-case is needed, and if anyone ever decides to extend those tests with a field with a fully-qualified type other than `HashMap<String, String>` then the tests will fail in a way that's probably going to be surprising.

On the other hand, it's actually not a realistic invariant that Rust -> JSON RPC -> Rust transform must always exactly reproduce the input Rust AST.  Codegen needs to fully-qualify types that would not normally be fully-qualified.  So another solution might be to run the decanonicalization pass on *both* `actual` and `expected`.  Now that I think of it, that seems like the better approach.  What do you think?

ahl replied:
I like the idea of running it on both actual and expected

anelson replied:
I just pushed https://github.com/oxidecomputer/typify/pull/705/commits/def217340862ddf63b4c1a3bbf00fc8b9af2f055 which does just that.  It's definitely better that way.

ahl reviewed (commented):
Thanks for this. The short of it seems to be to prefix std types with their full path. Any other subtlety beyond the testing kludge?

anelson commented:
> Thanks for this. The short of it seems to be to prefix std types with their full path. Any other subtlety beyond the testing kludge?

Yes that's pretty much the whole fix.  A few words to describe, and only +28,792 −13,699 lines to implement ;)

ahl commented on typify-impl/src/test_util.rs:
```diff
@@ -81,6 +88,10 @@ fn validate_output_impl<T: JsonSchema + Schema>(ignore_variant_names: bool) {
+    let actual = decanonicalize_std_types(actual);
+
     // Make sure they match.
     if let Err(err) = expected.syn_cmp(&actual, ignore_variant_names) {
```
would another approach be to change the implementation of `impl SynCompare for TypePath` to be fuzzier i.e. to identify `::std::...::Foo` as equivalent to `Foo`?

Alternatively, I'm inclined towards changing the tests to just use the fully qualified type name.

What do you think of these options?

ahl replied:
> Yes I suppose you could change the `SynCompare` impl to strip path components any time it sees `::std::` before the comparison. My personal preference is to not do that, since it's more spooky action at a distance, while this impl is very explicit about what's happening.

Fair enough.

> That would certainly work, but if anyone adds new test cases will they remember that the types have to be fully-qualified, and if they forget will the resulting test failure make it obvious what their mistake was?

I think so.

> It's your codebase, your call, but speaking in terms of my personal preference, I prefer the explicit decanonicalization approach.

Let's go with your approach.

ahl commented on typify-impl/src/type_entry.rs:
```diff
@@ -1480,17 +1486,18 @@ impl TypeEntry {
-                                        e.to_string(),
-                                    )
-                                })
+                            Self::try_from(
```
I'd prefer we keep this as `parse()` -- I have some ambivalence about the use of To/From<String> in these cases.

Are there other instances (I may have missed) where there's a functional change?

anelson replied:
No that should be the only one.  I've reverted this in a subsequent commit.

ahl reviewed (commented):
- how do we handle testing
- are there other functional changes (and if so, can we revert?)

model-context-protocol.json is large; is it adding value beyond what we get from rust-collisions?

Can you please take a look at rustdoc before and after? I don't recall if we'll see simplified or fully qualified types. The latter would be fine... but perhaps distracting / ugly.

Thanks.

anelson commented:
> * how do we handle testing

Sorry, I don't understand what you're asking.  I added two additional schema test cases to exercise the problem and verify the fix.  They produce code that compiles and passes all existing tests for correctness.  Is there some additional testing you have in mind?

> * are there other functional changes (and if so, can we revert?)

I will revert the change to the `Deserialize` impl for `String` types.  There are no other such changes that I can see.

> model-context-protocol.json is large; is it adding value beyond what we get from rust-collisions?

It isn't needed to reproduce the problem with types named `Result`, no.  I added it initially to reproduce that bug, but I kept it there since it seems useful to have as a test case another real-world JSON schema, in case of regressions.  You already have the Github and Vega schemas, so I figured there was precedent.  I can remove it if you disagree.

> Can you please take a look at rustdoc before and after? I don't recall if we'll see simplified or fully qualified types. The latter would be fine... but perhaps distracting / ugly.

Sure.

I ran `cargo doc -p example-build` from this branch with my codegen changes.

This is the doc for `example_build::builder::Veggie`:

<img width="1226" alt="example_build::builder::Veggie docs" src="https://github.com/user-attachments/assets/e85d4854-b7b5-4b94-a4d4-423d1be15de9">

Looks fine to me.  The fully-qualified types are simplified in the output.  And yet the actual codegen output is still using the canonicalized output:

<img width="821" alt="example_build::builder::Veggie Rust struct" src="https://github.com/user-attachments/assets/740d1058-4102-4d7d-9958-ff42300be0df">

ahl commented:
> > * how do we handle testing
> 
> Sorry, I don't understand what you're asking. I added two additional schema test cases to exercise the problem and verify the fix. They produce code that compiles and passes all existing tests for correctness. Is there some additional testing you have in mind?

I should have been clearer; I was summarizing my code review feedback.

> I will revert the change to the `Deserialize` impl for `String` types. There are no other such changes that I can see.

Thanks!

> > model-context-protocol.json is large; is it adding value beyond what we get from rust-collisions?
> 
> It isn't needed to reproduce the problem with types named `Result`, no. I added it initially to reproduce that bug, but I kept it there since it seems useful to have as a test case another real-world JSON schema, in case of regressions. You already have the Github and Vega schemas, so I figured there was precedent. I can remove it if you disagree.

I'm just a little reluctant to extend testing time (etc.) if there's not a ton of value; please think it over and do what you think is best.

> > Can you please take a look at rustdoc before and after? I don't recall if we'll see simplified or fully qualified types. The latter would be fine... but perhaps distracting / ugly.
> 
> Sure.
> 
> I ran `cargo doc -p example-build` from this branch with my codegen changes.

Thanks for checking on that!

anelson commented:
I just noticed in my comment https://github.com/oxidecomputer/typify/pull/705#issuecomment-2512300833, I just happened to pick at random one of the generated builder structs, and there is a `String` that is not canonicalized!

I looked into it further, it appears that `schemas.rs` never enabled generating struct builders!  I've just made another change to enable struct builders for all schemas.  As soon as I did that, the tests started to fail when generating builders for my cursed schema, so then I fixed the struct builder generator output to properly canonicalize `String`.

Note that as a result of this, the generated code for the test case `merged-schemas` compiles with warnings:

```

WARNINGS:
┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈
warning: unused variable: `value`
    --> tests/schemas/merged-schemas.rs:1825:13
     |
1825 |             value: Unsatisfiable2,
     |             ^^^^^ help: if this is intentional, prefix it with an underscore: `_value`
     |
     = note: `#[warn(unused_variables)]` on by default

warning: unused variable: `value`
    --> tests/schemas/merged-schemas.rs:1831:17
     |
1831 |         fn from(value: super::Unsatisfiable2) -> Self {
     |                 ^^^^^ help: if this is intentional, prefix it with an underscore: `_value`

warning: unused variable: `value`
    --> tests/schemas/merged-schemas.rs:1846:13
     |
1846 |             value: Unsatisfiable3,
     |             ^^^^^ help: if this is intentional, prefix it with an underscore: `_value`

warning: unused variable: `value`
    --> tests/schemas/merged-schemas.rs:1852:17
     |
1852 |         fn from(value: super::Unsatisfiable3) -> Self {
     |                 ^^^^^ help: if this is intentional, prefix it with an underscore: `_value`

warning: 4 warnings emitted
┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈

STDERR:
┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈
warning: unused variable: `value`
    --> /Users/rupert/sources/typify/typify/tests/schemas/merged-schemas.rs:1825:13
     |
1825 |             value: Unsatisfiable2,
     |             ^^^^^ help: if this is intentional, prefix it with an underscore: `_value`
     |
     = note: `#[warn(unused_variables)]` on by default

warning: unused variable: `value`
    --> /Users/rupert/sources/typify/typify/tests/schemas/merged-schemas.rs:1831:17
     |
1831 |         fn from(value: super::Unsatisfiable2) -> Self {
     |                 ^^^^^ help: if this is intentional, prefix it with an underscore: `_value`

warning: unused variable: `value`
    --> /Users/rupert/sources/typify/typify/tests/schemas/merged-schemas.rs:1846:13
     |
1846 |             value: Unsatisfiable3,
     |             ^^^^^ help: if this is intentional, prefix it with an underscore: `_value`

warning: unused variable: `value`
    --> /Users/rupert/sources/typify/typify/tests/schemas/merged-schemas.rs:1852:17
     |
1852 |         fn from(value: super::Unsatisfiable3) -> Self {
     |                 ^^^^^ help: if this is intentional, prefix it with an underscore: `_value`
┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈
```

I'm not sure if this is considered acceptable for the generated code or not.  It does not appear to be caused by any of my changes, it's just coming to light now that I've enabled generating builder structs in the test case.  To confirm this, I checked out `main`, added `.with_struct_builder(true)` to the `schemas.rs` test, and re-ran the tests; I get the same warnings.  It seems like it's an issue outside the scope of this PR but maybe you disagree.

anelson commented:
> I'm just a little reluctant to extend testing time (etc.) if there's not a ton of value; please think it over and do what you think is best.

I did some testing (all on my local M1 Pro MBP).

## `main` branch (without any of my test changes)

Right after a `cargo clean`:

```
cargo test -p typify  136.85s user 17.49s system 357% cpu 43.130 total
```

Re-running `cargo test` (so no rebuild, just re-running the already-built tests):

```
cargo test -p typify  4.89s user 3.62s system 73% cpu 11.540 total
```

## this branch with my changes

Right after a `cargo clean`:

```
cargo test -p typify  140.43s user 17.88s system 279% cpu 56.583 total
```

Re-running `cargo test` (so no rebuild, just re-running the already-built tests):

```
cargo test -p typify  5.67s user 3.98s system 73% cpu 13.050 total
```

## this branch with my changes, but with the MCP schema removed

Right after a `cargo clean`:

```
cargo test -p typify  135.65s user 17.23s system 348% cpu 43.851 total
```

Re-running `cargo test` (so no rebuild, just re-running the already-built tests):

```
cargo test -p typify  5.30s user 4.03s system 72% cpu 12.929 total
```

Key findings:

- The changes increase clean build time by 31.2% (+13.5s)
- The MCP schema accounts for most of this increase (12.7s)
- Test re-runs are marginally slower (+13.1%, or 1.5s)

So I don't think the MCP schema is providing enough value to justify the increase in build/test time.  I'll remove it.

anelson commented on typify/tests/schemas.rs:
```diff
@@ -64,7 +64,8 @@ fn validate_schema(path: std::path::PathBuf) -> Result<(), Box<dyn Error>> {
                 None,
-            ),
+            )
+            .with_struct_builder(true),
```
See, see my comment https://github.com/oxidecomputer/typify/pull/705#issuecomment-2512335987 for a detailed explanation.  There was one place left in the codegen where `String` was used, and all of the tests passed because that one place was in the codegen for optional struct builders.  So I enabled that for all of the schema tests, the tests started to fail, then I fixed that one remaining use of a simple `String` type and verified that the tests passed.  I left this in place because it seems like these tests are meant to exercise the code generation in its entirety.

ahl commented on typify-impl/src/test_util.rs:
```diff
@@ -104,18 +104,16 @@ fn validate_output_impl<T: JsonSchema + Schema>(ignore_variant_names: bool) {
-/// that type, and expects that the Rust AST for the source type and the generated type are exactly
-/// the same.
+/// this complicates the test cases that validate that a round-trip from Rust to JSON back to Rust
+/// are exactly the same.
```
wrap, remove comma

anelson replied:
I rewrapped and removed the extraneous comma.  

I rebased against upstream `main` to resolve the conflict with `Cargo.toml`.

Hopefully this is good to go now...

ahl reviewed (approved):
one fix, merge Cargo.toml, and we'll merge

ahl commented:
Thanks so much for this contribution! Greatly appreciated!
